content/en/docs/hertz/tutorials/advanced-exten/protocol.md

```diff
@@ -0,0 +1,166 @@
+
+```go
+type ServerFactory interface {
+   New(core Core) (server protocol.Server, err error)
+}
+
+// Core is the core interface that promises to be provided for the protocol layer extensions
+type Core interface {
+   // IsRunning Check whether engine is running or not
+   IsRunning() bool
+   // A RequestContext pool ready for protocol server impl
+   GetCtxPool() *sync.Pool
+   // Business logic entrance
+   // After pre-read works, protocol server may call this method
+   // to introduce the middlewares and handlers
+   ServeHTTP(c context.Context, ctx *app.RequestContext)
+   // GetTracer for tracing requirement
+   GetTracer() tracer.Controller
+}
+```
+
+A custom server only needs to implement a protocol server generation factory according to the above interface. The Core in the parameters actually includes the introduction of upper-layer logic interaction and the specific implementation of other core application layer interfaces. When initializing a custom server, normally you only need to save the Core to the server. When you need to transfer to the business logic, you can guide the process to the application layer processing logic (Route, Middleware, Logic Handler) through the Core. When the business logic is executed and returned, you can further packet writeback is performed based on the business data.
+
+```go
+type myServer struct{
+    suite.Core
+    xxx
+}
+
+func (s *myServer) Serve (c context.Context, conn network.Conn) error{
+    // protocol parsing
+	...
+    Core.ServeHTTP(c, ctx)
+    // write data back
+	...
+}
+```
+
+So far, a custom protocol layer server has been developed.
+
+### Register custom protocol server with Hertz
+
+After completing the development of the server generation factory according to the above interface, it is very easy to load it into Hertz. Hertz's core engine naturally provides an interface for registering a custom protocol server:
+
+```go
+func (engine *Engine) AddProtocol(protocol string, factory suite.ServerFactory) {
+   engine.protocolSuite.Add(protocol, factory)
+}
+```
+
+It is only necessary to register the user's custom server generation factory with the engine according to the parameters specified by the interface. But it is worth noting that the protocol (string) registered here actually corresponds to the protocol negotiation key in ALPN (Application-Layer Protocol Negotiation), so if you want to access a custom protocol server through ALPN , directly specify the key as the corresponding key during ALPN negotiation. Currently, Hertz integrates an HTTP1 protocol server by default (the corresponding key is "http/1.1"). If you need to customize the HTTP1 protocol processing logic, you can directly specify the key as "http/1.1" when `AddProtocol` Complete coverage.
```

Comment:
... you can directly specify the key as "http/1.1" with `AddProtocol` to overwrite.